app/src/org/commcare/android/database/app/models/UserKeyRecord.java

```diff
@@ -235,4 +344,86 @@ public boolean isCurrentlyValid() {
                 (validTo == null ||
                         (validTo.getTime() != Long.MAX_VALUE && validTo.after(today))));
     }
+
+    public static UserKeyRecord getCurrentValidRecordByPassword(CommCareApp app, String username,
+                                                                String pw, boolean acceptExpired) {
+        return getCurrentValidRecord(app, username, pw, null, acceptExpired);
+    }
+
+    public static UserKeyRecord getCurrentValidRecordByPin(CommCareApp app, String username,
+                                                           String pin, boolean acceptExpired) {
+        return getCurrentValidRecord(app, username, null, pin, acceptExpired);
+    }
+
+    public static UserKeyRecord getMatchingPrimedRecord(CommCareApp app, String username) {
+        SqlStorage<UserKeyRecord> storage = app.getStorage(UserKeyRecord.class);
+        for (UserKeyRecord ukr : storage.getRecordsForValue(UserKeyRecord.META_USERNAME, username)) {
+            if (ukr.isPrimedForNextLogin()) {
+                return ukr;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * @return The user record that matches the given username/password or username/pin combo.
+     * Null if not found or user record validity date is expired.
+     */
+    private static UserKeyRecord getCurrentValidRecord(CommCareApp app, String username, String pw,
+                                                       String pin, boolean acceptExpired) {
+        UserKeyRecord invalidRecord = null;
+        SqlStorage<UserKeyRecord> storage = app.getStorage(UserKeyRecord.class);
+
+        for (UserKeyRecord ukr : storage.getRecordsForValue(UserKeyRecord.META_USERNAME, username)) {
+            if (ukr.isPasswordOrPinValid(pw, pin)) {
+                if (ukr.isCurrentlyValid()) {
+                    return ukr;
+                } else {
+                    invalidRecord = ukr;
+                }
+            }
+        }
+
+        if (acceptExpired) {
+            return invalidRecord;
+        }
+        return null;
+    }
+
+    public void setPrimedPassword(String unhashedPassword) {
+        this.rememberedPassword = unhashedPassword;
+    }
+
+    public boolean isPrimedForNextLogin() {
+        return !"".equals(rememberedPassword);
+    }
+
+    public String getPrimedPassword() {
+        return this.rememberedPassword;
+    }
+
+    public void clearPrimedPassword() {
+        this.rememberedPassword = "";
+    }
```

Comment:
What are you looking to be added? They're pretty simple methods with descriptive names, not sure what I would comment. I added a comment to the rememberedPassword field as you suggested, which might provide the clarification you're looking for on all of these.